sling/nlp/document/lex.cc

```diff
@@ -184,33 +212,52 @@ string ToLex(const Document &document) {
       }
     }
 
+    // Add style begin.
+    if (style != 0) {
+      int begin_style = style & BEGIN_STYLE;
+      OutputStyle(begin_style, &output);
+      styles |= begin_style << 1;
+    }
+
     // Add span open brackets.
     Span *span = document.GetSpanAt(token.index());
     for (Span *s = span; s != nullptr; s = s->parent()) {
       if (s->begin() == token.index()) output.WriteChar('[');
     }
 
-    // Add token word.
-    output.Write(token.word());
+    // Add token word. Escape reserved characters.
+    for (char c : token.word()) {
+      switch (c) {
+        case '{': case '[': c = '('; break;
+        case '}': case ']': c = ')'; break;
+        case '|': c = '!'; break;
+      }
+      output.WriteChar(c);
+    }
 
     // Add span close brackets.
     for (Span *s = span; s != nullptr; s = s->parent()) {
       if (s->end() == token.index() + 1) {
         bool first = true;
         s->AllEvoked(&evoked);
         for (Handle frame : evoked) {
-          if (first) output.WriteChar('|');
+          output.WriteChar(first ? '|' : ' ');
           first = false;
-          printer.Print(frame);
+          printer.PrintReference(frame);
         }
         output.WriteChar(']');
       }
     }
   }
 
+  // Terminate remaining styles.
+  if (styles != 0) {
+    OutputStyle(styles, &output);
```

Comment:
is this correct? let's say there is an unterminated bold tag, so styles = BOLD_BEGIN, so OutputStyle() would erroneously open a <b> tag here, right?

Reply:
Maybe this is a little subtle, but the `styles` variable holds all the end tags that we have not seen yet. Line 219 (`styles |= begin_style << 1;`) uses the fact that the end tag mask is the begin tag mask shifted one to the left.